titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultDirectKubeApiServerIntegrator.java

```diff
@@ -218,6 +214,62 @@ public String resolveReasonCode(Throwable cause) {
         return podCreateErrorToReasonCodeResolver.resolveReasonCode(cause);
     }
 
+    private Mono<V1PersistentVolume> launchPersistentVolume(Job<?> job, Task task, V1Pod v1Pod, CoreV1Api coreV1Api) {
+        return Mono.defer(() -> {
+            Stopwatch timer = Stopwatch.createStarted();
+
+            Optional<V1Volume> optionalV1Volume = CollectionsExt.nonNull(v1Pod.getSpec().getVolumes())
+                    .stream()
+                    .filter(v1Volume -> null != v1Volume.getAwsElasticBlockStore())
+                    .findFirst();
+            if (!optionalV1Volume.isPresent()) {
+                return Mono.empty();
+            }
+            V1Volume v1Volume = optionalV1Volume.get();
+            V1PersistentVolume v1PersistentVolume = KubeModelConverters.toV1PersistentVolume(job, task, v1Volume);
+
+            logger.info("Creating persistent volume {} for task {}", v1PersistentVolume, task.getId());
+
+            try {
+                coreV1Api.createPersistentVolume(v1PersistentVolume, null, null, null);
+                logger.info("Created persistent volume {} in {}", v1PersistentVolume, timer.elapsed(TimeUnit.MILLISECONDS));
+                metrics.persistentVolumeCreateSuccess(timer.elapsed(TimeUnit.MILLISECONDS));
+            } catch (ApiException apiException) {
+                if (isEbsVolumeConflictException(apiException)) {
+                    logger.info("Persistent volume already exists {}", v1PersistentVolume);
+                } else {
+                    logger.error("Unable to create persistent volume {}, error: {}", v1PersistentVolume, apiException);
+                    metrics.persistentVolumeCreateError(apiException, timer.elapsed(TimeUnit.MILLISECONDS));
```

Comment:
Is there a value in tracking errors by fsType and AZ?

Reply:
That's a good question. What we capture here is the same level of error detail we capture for pods. Specific fields that caused a create failure we buried in the Kube error message string.